libyul/YulName.h

```diff
@@ -0,0 +1,230 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#pragma once
+
+#include <libyul/ControlFlowSideEffects.h>
+#include <libyul/SideEffects.h>
+
+#include <fmt/format.h>
+
+#include <unordered_map>
+#include <memory>
+#include <vector>
+#include <string>
+#include <optional>
+#include <functional>
+#include <numeric>
+#include <map>
+
+namespace solidity::yul
+{
+
+enum class LiteralKind;
+struct Dialect;
+struct Block;
+struct BuiltinFunction;
+
+/**
+ * Yul Name repository.
+ *
+ * A database of numerical identifiers for Yul nodes in an AST (``YulName``s). Identifiers can be
+ *
+ *  - 'defined', i.e., they are equipped with a string label, which can be retrieved by a call to ``labelOf(yulName)``;
+ *  - 'derived', i.e., they don't possess a string label but have a parent yul name which is either also derived or
+ *    defined. All dependency chains of derived labels terminate in a defined label.
+ *
+ * Such derived identifiers can be introduced, e.g., during certain optimization steps. If the AST (or segments thereof)
+ * should be printed and/or the string label of identifiers should be retrieved which are still in derived state,
+ * a call to ``generateLabels(identifiers)`` changes the status of all derived labels to defined and generates
+ * unique labels for all identifiers provided to the method. The generated labels are based on their parents.
+ *
+ * The second purpose of the YulNameRepository is acting as a proxy for a ``yul::Dialect``. Builtins as well
+ * as types are mapped to entries in the repository and can be queried using corresponding methods.
+ *
+ * There are two special cases:
+ *  - verbatim functions are builtin functions which are already variable, therefore they are derived from a common
+ *    ancestor ``PredefinedHandles.verbatim`` but don't need a label generated for them, as the label is already
+ *    defined by their number of inputs and outputs
+ *  - GHOST variables are used for specific transformations in the ``ControlFlowGraph`` and can be added
+ *    using ``addGhost``. These always fall under the 'defined' category.
+ */
+class YulNameRepository
+{
+public:
+	using YulName = size_t;
```

Comment:
Let's use a type that has a consistent range between platforms:
```suggestion
	using YulName = uint64_t;
```